### PR TITLE
rpc: Faster getblock API

### PR DESCRIPTION
**Pull request by AaronDewes**
Previously, `cs_main` had to be locked during `GetBlockChecked` in the `getblock` rpc call. This made parallel `getblock` calls slower, for example in Fulcrum.

By moving the pruned check out of `GetBlockChecked` into a separate function, `cs_main` is no longer locked during `ReadBlockFromDisk`.

It still locks `cs_main` if pruning is enabled to prevent a race condition (#13903), so the performance improvement will only be noticeable without pruning.

### CODE REVIEW

**Review by vincenzopalazzo — COMMENTED** [review]
There is some way to ensure that this is a speed improvement that justify this refactoring?

**Comment by AaronDewes** [comment]
> There is some way to ensure that this is a speed improvement that justify this refactoring?

Yes, I'm running a benchmark right now and will share the results. A good way to test is also trying to sync [Fulcrum](https://github.com/cculianu/Fulcrum), an Electrum server for a more real-world scenario where getblock is used a lot.

**Comment by AaronDewes** [comment]
I haven't found a good way to benchmark it except trying to sync Fulcrum and viewing the logs. Without this patch, it often exceeded the RPC work queue depth and was very slow. With this patch, it was much faster and didn't do that.

**Comment by laanwj** [comment]
Concept ACK on doing the block read from disk without locking, when pruning is not enabled. This seems quite low-hanging fruit for optimization. Assuming, there can't be a race condition between the block being *written* to disk and reading it?

**Comment by AaronDewes** [comment]
I don't think thereis, because the check if the block exists (`LookupBlockIndex`) happens with `cs_main` locked. So if it's not written before the lock is released, the reading won't happen.

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -565,14 +565,18 @@ static RPCHelpMan getblockheader()
     };
 }
 
-static CBlock GetBlockChecked(BlockManager& blockman, const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
+static void EnsureNotPruned(BlockManager& blockman, const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
Comment (aureleoules) [inline src/rpc/blockchain.cpp:568]: nit: i think it's clearer
```suggestion
static void EnsureBlockNotPruned(BlockManager& blockman, const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```

**Review by aureleoules — COMMENTED** [review]
ACK c46d05144d1440889251cb01e6ae8361ce4381a8.

I'm not sure how to benchmark this either so I used the poor's man benchmark tool `time`:
`time seq 10000 | xargs -i -P 15 ./src/bitcoin-cli -testnet getblock 000000000000001b3a9aeab5659c3e13dbc32d771a8b4f86aeac529d51a12039`
This executes 10,000 `getblock` requests in parallel batches of `15` on testnet.

### Master
46.48s user 21.84s system 395% cpu 17.277 total
46.44s user 21.75s system 419% cpu 16.254 total

### PR
45.73s user 21.27s system 425% cpu 15.740 total
45.74s user 21.40s system 422% cpu 15.894 total

The results are slightly faster than master but they're consistent.

**Comment by aureleoules** [comment]
You'll need to squash the new commit https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

**Review by aureleoules — APPROVED** [review]
ACK 599f2608c60b72f6ea0b6f109f0d16b3e6afdcb5

**Comment by luke-jr** [comment]
What if pruning occurs for the first time, after the lock is released, and before it is read?

IMO a better approach would be to use a temporary prune lock. This would extend the benefit to pruned nodes, and could possibly (in another PR) be used to safely re-fetch a pruned block from another peer.

**Comment by andrewtoth** [comment]
Sorry I wasn't aware of this PR until drahtbot alerted me to conflicts. You can bench with ApacheBench. Run
```
ab -p data.json -n 10000 -c 16 -A user:password "http://127.0.0.1:8332/"
```
where `data.json` is 
```
{"jsonrpc": "1.0", "id": "curltest", "method": "getblock", "params": ["0000000000000000000592a974b1b9f087cb77628bb4a097d5c2c11b3476a58e"]}
```
and the `-c` arg is the number of `rpcthread` you have in your `bitcoin.conf`. You should set `rpcthread` to the number of CPU threads on your bench machine. Also try with just `-c 1` to make sure there's no regression on single thread.

> What if pruning occurs for the first time, after the lock is released, and before it is read?

If the file has not yet been opened, it will fail and throw an RPC error similar to if it is pruned. If the file has already been opened, then it should succeed and remove the file once the file is closed on linux. On Windows it could cause the files to not be pruned https://en.cppreference.com/w/cpp/io/c/remove. See https://github.com/bitcoin/bitcoin/pull/26316.

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -722,7 +727,15 @@ static RPCHelpMan getblock()
             throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
         }
 
-        block = GetBlockChecked(chainman.m_blockman, pblockindex);
+        havePruned = chainman.m_blockman.m_have_pruned;
+        if(havePruned) {
+            EnsureBlockNotPruned(chainman.m_blockman, pblockindex);
+            block = GetBlockChecked(pblockindex);
```
Comment (MarcoFalke) [inline src/rpc/blockchain.cpp:733]: Is there any value in keeping the cs_main lock here. In the very rare case where the block could be deleted after `EnsureBlockNotPruned` but before `GetBlockChecked`  without the lock, it seems fine to just error out as well. The only difference would the the error message.

This is done in https://github.com/bitcoin/bitcoin/pull/26308

**Comment by andrewtoth** [comment]
@AaronDewes I think #26308 accomplishes this. I syncing Fulcrum on master faster now?

**Comment by AaronDewes** [comment]
Yes, it seems a lot faster, thank you!

**Pull request closed without merging**